src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperDescriptorFactory.cs

```diff
@@ -750,5 +733,328 @@ private static string ToHtmlCase(string name)
         {
             return HtmlCaseRegex.Replace(name, HtmlCaseRegexReplacement).ToLowerInvariant();
         }
+
+        // Internal for testing
+        internal class RequiredAttributeParser
+        {
+            private int _index;
+            private string _requiredAttributes;
+
+            public RequiredAttributeParser(string requiredAttributes)
+            {
+                _requiredAttributes = requiredAttributes;
+            }
+
+            private char Current => _requiredAttributes[_index];
+
+            private bool AtEnd => _index >= _requiredAttributes.Length;
+
+            public bool TryParse(
+                ErrorSink errorSink,
+                out IEnumerable<TagHelperRequiredAttributeDescriptor> requiredAttributeDescriptors)
+            {
+                if (_requiredAttributes == null)
+                {
+                    requiredAttributeDescriptors = Enumerable.Empty<TagHelperRequiredAttributeDescriptor>();
+                    return true;
+                }
+
+                requiredAttributeDescriptors = null;
+                var descriptors = new List<TagHelperRequiredAttributeDescriptor>();
+
+                while (!AtEnd)
+                {
+                    PassOptionalWhitespace();
+
+                    TagHelperRequiredAttributeDescriptor descriptor;
+                    if (At('['))
+                    {
+                        descriptor = ParseCSSSelector(errorSink);
+                    }
+                    else
+                    {
+                        descriptor = ParsePlainSelector(errorSink);
+                    }
+
+                    if (descriptor == null)
+                    {
+                        // Failed to create the descriptor due to an invalid required attribute.
+                        return false;
+                    }
+                    else
+                    {
+                        descriptors.Add(descriptor);
+                    }
+
+                    PassOptionalWhitespace();
+
+                    if (At(','))
+                    {
+                        Next();
+                        PassOptionalWhitespace();
+                    }
+                    else if (!AtEnd)
+                    {
+                        errorSink.OnError(SourceLocation.Zero, $"TODO: Unknown character '{Current}'. Separate required attributes with commas.", 0);
+                        return false;
+                    }
+                }
+
+                requiredAttributeDescriptors = descriptors;
+                return true;
+            }
+
+            private TagHelperRequiredAttributeDescriptor ParsePlainSelector(ErrorSink errorSink)
+            {
+                var nextSeparator = _requiredAttributes.IndexOf(',', _index);
```

Comment:
Disagree because the rest of this method is mostly about characters _after_ the name. If `afterNameIndex` is set correctly and not `-1`, easy to see whether `_requiredAttributres[afterNameIndex] == RequiredAttributeWildcardSuffix`. And there's no reason for this method to consume whitespace.